creator-node/src/services/stateMachineManager/monitoring/StateMonitoringQueue.js

```diff
@@ -0,0 +1,254 @@
+      startTime: Date.now(),
+      lastProcessedUserId,
+      discoveryNodeEndpoint,
+      moduloBase,
+      currentModuloSlice: (currentModuloSlice + 1) % moduloBase
+    })
+  }
+
+  /**
+   * Enqueues a job that picks up where the previous failed job left off.
+   * @param queue the StateMonitoringQueue to enqueue a job
+   * @param failedJob the jobData for the previous job that failed
+   */
+  enqueueJobAfterFailure(queue, failedJob) {
+    const {
+      data: {
+        lastProcessedUserId,
+        discoveryNodeEndpoint,
+        moduloBase,
+        currentModuloSlice
+      }
+    } = failedJob
+
+    queue.add({
+      startTime: Date.now(),
+      lastProcessedUserId,
+      discoveryNodeEndpoint,
+      moduloBase,
+      currentModuloSlice: (currentModuloSlice + 1) % moduloBase
+    })
+  }
+
+  /**
+   * Registers the logic that gets executed to process each new job from the queue.
+   * @param {*} queue the StateMonitoringQueue to consume jobs from
+   */
+  registerQueueConsumer(queue) {
+    // Initialize queue job processor (aka consumer)
+    queue.process(1 /** concurrency */, async (job) => {
+      const {
+        id: jobId,
+        data: {
+          lastProcessedUserId,
+          discoveryNodeEndpoint,
+          moduloBase,
+          currentModuloSlice
+        }
+      } = job
+
+      try {
+        this.log(`New job details: jobId=${jobId}, job=${JSON.stringify(job)}`)
+      } catch (e) {
+        this.logError(`Failed to log details for jobId=${jobId}: ${e}`)
+      }
+
+      // Default results of this job will be passed to the next job, so default to failure
+      let result = {
+        lastProcessedUserId,
+        jobFailed: true,
+        moduloBase,
+        currentModuloSlice
+      }
+      try {
+        // TODO: Wire up metrics
+        // await redis.set('stateMachineQueueLatestJobStart', Date.now())
+        result = await processStateMonitoringJob(
+          jobId,
+          lastProcessedUserId,
+          discoveryNodeEndpoint,
+          moduloBase,
+          currentModuloSlice
+        )
+        // TODO: Wire up metrics
+        // await redis.set('stateMachineQueueLatestJobSuccess', Date.now())
+      } catch (e) {
+        this.logError(`Error processing jobId ${jobId}: ${e}`)
+        console.log(e.stack)
+      }
+
+      return result
+    })
+  }
+
+  /**
+   * Clears the queue and adds a job that will start processing users
+   * starting from a random userId. Future jobs are added to the queue as a
+   * result of this initial job succeeding or failing to complete.
+   * @param {Object} queue the StateMonitoringQueue to consume jobs from
+   * @param {string} discoveryNodeEndpoint the IP address or URL of a Discovery Node
+   * @param {number} moduloBase (DEPRECATED)
+   */
+  async startQueue(queue, discoveryNodeEndpoint, moduloBase) {
+    // Clear any old state if redis was running but the rest of the server restarted
+    await queue.obliterate({ force: true })
```

Comment:
separate point, but would love your input on this pattern - obliterate is a very heavy weight op, idk if we need to do it each time

it was one of my attempts at restarting the queue but at this point i think we're confident it was unrelated right

Reply:
synced offline and was going to change to clean+empty but after some digging I'm thinking maybe we should use obliterate just in case. [this comment](https://github.com/OptimalBits/bull/issues/709#issuecomment-417870998) about repeatable jobs makes me wonder if we ever change this to a repeatable job, it might pollute our redis and possibly cause unexpected behavior. avoiding that pollution/unexpected behavior is worth the loss of history on restarts imo. wdyt?

Reply:
thanks for digging into this - your point about losing job history is a big one. don't need to block merge since this is unrelated to refactor, but we should not forget about this

Reply:
synced offline and decided to keep it as obliterate to err on the side of safety in exchange for poorer log history during server restarts. if there are a lot of jobs to obliterate, this could make Content Node startup slow down a bit so we'll keep an eye out for that